README.md
# README #

Octopress plugin to display tag clouds. Based on https://gist.github.com/1577100 by tokkonopapa

Version: 0.4


## Getting Started #

This won't work without a plugin to generate some tag pages. See https://github.com/robbyedwards/octopress-tag-pages for a plugin to take care of this.

Put `tag_cloud.rb` in the `/plugins` folder. Then add `{% tag_cloud %}` where the tag cloud should appear. A couple of sample files are included to demonstrate basic usage, and several options are provided to control the appearance of the tag cloud.


## Usage #

Defines a `tag_cloud` tag that is rendered by Liquid into a tag cloud:

    <div class="cloud">
        {% tag_cloud %}
    </div>

Available _config.yml settings :

- `tag_dir`: The subfolder to build tag pages in (default is `tags`).

The tag cloud itself is a collection of anchor tags, styled dynamically with the `font-size` CSS property. The range of values, and unit to use for `font-size` can be specified with a very simple syntax:

    {% tag_cloud font-size: 16 - 28px %}

The output is automatically formatted to use the same number of decimal places as used in the argument:

    {% tag_cloud font-size: 0.8  - 1.8em  %}  # => 1
    {% tag_cloud font-size: 0.80 - 1.80em %}  # => 2

Tags that have been used less than a certain number of times can be filtered out from the tag cloud with the optional `threshold` parameter:

    {% tag_cloud threshold: 2 %}

The ouptput can be limited to a certain number of the most frequently occurring tags when the sort method is alphabetical or frequency; i.e., `{% tag_cloud limit: 8 %}` => results in a tag cloud with the eight most frequent tags sorted alphabetically.

When the sort method is random then output is randomly sorted and limited to `@limit` items. Setting the `limit` parameter to zero (the default) will display all tags.

The output can be sorted in three ways:

    {% tag_cloud sort: freq|rand|alpha %}

- `freq` will sort by the frequency of the tag`s occurrence
- `rand` will sort randomly each time the cloud is generated
- `alpha` (the default) will sort the tags in alphabetical order

An optional `order` parameter can also be set, which will change the sort order:

    {% tag_cloud sort: freq|rand|alpha [asc|desc] %}

By default, alphabetical sorts will be asc, and frequency sorts will be desc. The `order` parameter has no effect on random sorts.

The `style` parameter provides a basic way to alter the html that is generated by the plugin:

    {% tag_cloud style: list|para [{:separator}] %}

- `list` wraps each anchor tag in an `<li>` tag for presenting the list as an (un)ordered list. The `{% tag_cloud %}` block needs to be enclosed in either a `<ol>` or a  `<ul>` block where it is included in the site's templates. `list` is the default option. The optional separator will have no effect when the `list` option is used
- `para` removes the enclosing `<li>` tags and separates each item with an optional separator (indicated by `:separator` above). By default the separator is a comma followed by a space. An alternate separator can be indicated by wrapping the new separator inside curly brackets ({}) following the `para` option, i.e., `{% tag_cloud style: para { &mdash; } %}`

Multiple parameters can be separated by a comma:

    {% tag_cloud font-size: 50 - 150%, threshold: 2 %}


## License #

Copyright (c) 2012 Robby Edwards, http://robbyedwards.com/
Licensed under the MIT license (http://www.opensource.org/licenses/mit-license.php)

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the ‘Software’), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED ‘AS IS’, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


CHANGELOG.md
# CHANGELOG #

## 0.4 #
- addresses a couple of issues:
  - [FIXED] When not using the `tag_dir` setting, links are now created with the default tag directory.
  - [ENHANCEMENT] Added stringex and changed to use #to_url to create the tag directory.

## 0.3 #
- added README

## 0.2 #
- fixed generated URLs

## 0.1 #
- initial version

plugins/tag_cloud.rb
# encoding: utf-8
#
# Octopress tag cloud generator
#
# Version: 0.3
#
# Copyright (c) 2012 Robby Edwards, http://robbyedwards.com/
# Licensed under the MIT license (http://www.opensource.org/licenses/mit-license.php)
#
# Octopress plugin to display tag clouds.
# Based on https://gist.github.com/1577100 by @tokkonopapa
#
# Defines a 'tag_cloud' tag that is rendered by Liquid into a tag cloud:
#
#     <div class='cloud'>
#         {% tag_cloud %}
#     </div>
#
# See README for installation and usage instructions.

require 'stringex'

module Jekyll

  class TagCloud < Liquid::Tag
    safe = true

    # tag cloud variables - these are setup in 'initialize'
    attr_reader :size_min, :size_max, :precision, :unit, :threshold, :limit, :sort, :order, :style, :separator

    def initialize(name, params, tokens)
      # initialize default values
      @size_min, @size_max, @precision, @unit = 70, 170, 0, '%'
      @threshold                              = 1
      @limit                                  = 0
      @sort, @order                           = 'alpha', 'asc'
      @order = 'desc' if @sort == 'freq'
      @style, @tag_before, @tag_after, @separator = 'list', '<li>', '</li>', ', '

      # process parameters
      @params = Hash[*params.split(/(?:: *)|(?:, *)/)]
      process_font_size(@params['font-size'])
      process_threshold(@params['threshold'])
      process_limit(@params['limit'])
      process_sort(@params['sort'])
      process_style(@params['style'])

      super
    end

    def render(context)
      # get the directory for the tag links
      dir = context.registers[:site].config['tag_dir'] || 'tags'

      # get an Array of [tag name, tag count] pairs
      count = context.registers[:site].tags.map do |name, posts|
        [name, posts.count] if posts.count >= threshold
      end

      # clear nils if any
      count.compact!

      # get the minimum, and maximum tag count
      min, max = count.map(&:last).minmax

      # map: [[tag name, tag count]] -> [[tag name, tag weight]]
      weighted = count.map do |name, count|
        # logarithmic distribution
        weight = (Math.log(count) - Math.log(min))/(Math.log(max) - Math.log(min))
        [name, weight]
      end

      # get the top @limit tag pairs when a limit is given, unless the sort method is random
      if @limit > 0 and @sort != 'rand'
        # sort the tag pairs by frequency in descending order
        weighted.sort! { |a,b| b[1] <=> a[1] }

        # then slice off the top @limit tag pairs
        weighted = weighted[0,@limit]
      end

      # sort the [tag name, tag weight] pairs
      case @sort
      when 'freq'
        if @order == 'asc'
          # sorts from least to most frequent
          weighted.sort! { |a,b| a[1] <=> b[1] }
        elsif @limit == 0
          # otherwise, sort from the most to least frequent
          weighted.sort! { |a,b| b[1] <=> a[1] }
        end
      when 'rand'
        weighted.sort_by! { rand }

        if @limit > 0
          # slice off the top @limit tag pairs
          weighted = weighted[0,@limit]
        end
      else
        if @order == 'desc'
          # sorts in reverse alphabetical order, i.e z to a
          weighted.sort! { |a,b| b <=> a }
        else
          # otherwise, sorts in alphabetical order, i.e a to z
          weighted.sort! { |a,b| a <=> b }
        end
      end

      html = ""

      # iterate over the weighted tag Array and create the tag items
      weighted.each_with_index do |tag, i|
        name, weight = tag
        size = size_min + ((size_max - size_min) * weight).to_f
        size = sprintf("%.#{@precision}f", size)
        slug = name.to_url
        @separator = "" if i == (weighted.size - 1)
        html << "#{@tag_before}<a style=\"font-size: #{size}#{unit}\" href=\"/#{dir}/#{slug}/\">#{name}</a>#{@separator}#{@tag_after}\n"
      end

      html
    end

    private

    def process_font_size(param)
      /(\d*\.{0,1}(\d*)) *- *(\d*\.{0,1}(\d*)) *(%|em|px)/.match(param) do |m|
        @size_min  = m[1].to_f
        @size_max  = m[3].to_f
        @precision = [m[2].size, m[4].size].max
        @unit      = m[5]
      end
    end

    def process_threshold(param)
      /\d*/.match(param) do |m|
        @threshold = m[0].to_i
      end
    end

    def process_limit(param)
      /\d*/.match(param) do |m|
        @limit = m[0].to_i
      end
    end

    def process_sort(param)
      /(freq|rand|alpha) *(asc|desc)?/.match(param) do |m|
        @sort  = m[1]
        @order = m[2]
      end
    end

    def process_style(param)
      /(list|para) *({(.*)})?/.match(param) do |m|
        @style     = m[1]
        @separator = m[3]
      end

      @tag_before = @tag_after = "" if @style == "para"
      @separator = "" if @style == "list"
    end

  end

end

Liquid::Template.register_tag('tag_cloud', Jekyll::TagCloud)
